# Exercise 1.1 for Basis set Optimization

In [14]:
!pwd

/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm/ex_1


In [15]:
%cd /home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm/

/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm


In [16]:
ls

ex_1/  EXERCISE-1.ipynb  H.psml  O.psml  water.fdf


In [3]:
!mkdir ex_1

In [4]:
!cp H.psml  O.psml  water.fdf ex_1/

In [17]:
!ls ex_1

0_NORMAL_EXIT		     h2o.bib	      h2o.STRUCT_OUT   O.ion
BASIS_ENTHALPY		     h2o.BONDS	      h2o.XV	       O.ion.nc
BASIS_HARRIS_ENTHALPY	     h2o.BONDS_FINAL  H.ion	       O.ion.xml
CLOCK			     h2o.DM	      H.ion.nc	       O.psml
fdf.20260106T143228.352.log  h2o.EIG	      H.ion.xml        OUTVARS.yml
fdf.20260106T143842.225.log  h2o.FA	      H.psml	       PARALLEL_DIST
FORCE_STRESS		     h2o.HSX	      INPUT_TMP.10943  water.fdf
h2o.alloc		     h2o.KP	      INPUT_TMP.84818  water.out
h2o.BASIS_ENTHALPY	     h2o.ORB_INDX     MESSAGES


In [18]:
%cd ex_1

/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm/ex_1


In [19]:
!siesta < water.fdf > water.out

Job completed


In [21]:
!vi water.out

7[>4;2m="water.out" 738 lines, 29142 bytesExecutable      : siesta
Version: 5.4.2
Architecture    : x86_64
Compiler version: GNU-14.3.0
Compiler flags  : -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-prrotector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem <prefix>/include  -I<prefix>/_build_env/include -fdebug-prefix-map=<prefix>/work=/usr/local/src/coonda/siesta-5.4.2 -fdebug-prefix-map=<prefix>=/usr/local/src/conda-prefix -I<preffix>/lib -fallow-argument-mismatch -O3
Parallelisations: none
NetCDF support
NetCDF-4 support
Lua support
DFT-D3 support

Runtime information:
* Directory : /home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optiimization/01-Eshift_SplitNorm/ex_1
* Running in serial mode.
>> Start of run:   6-JAN-2026  14:40:06************************  WELCOME TO SIESTA  *Type  :qa  and press <Enter> to exit Vim

In [24]:
!grep -F "E + p_basis" water.out


    (Free)E + p_basis*V_orbitals       =        -482.064557 eV


In [25]:
!cat h2o.BASIS_ENTHALPY 

 Basis enthalpy [eV] :   -482.06455744757909     
 Basis Harris enthalpy [eV] :   -482.06455746952162     
 Free energy [eV] :   -482.13276932086330     
 Harris energy [eV] :   -482.13276934280583     
 Orbital volume [Ang**3] :    546.43734768656418     
 Average basis pressure [eV/Ang**3] :    1.2483018148921527E-004
 Average basis pressure [GPa] :    2.0000000000000000E-002


#  Exercise 1.2 Finding E + p_basis(Basis enthalpy) at different energyshift

In [29]:
%cd /home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optiimization/01-Eshift_SplitNorm/


[Errno 2] No such file or directory: '/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optiimization/01-Eshift_SplitNorm/'
/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm/ex_1


In [30]:
!pwd

/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm/ex_1


In [31]:
%cd ..

/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm


In [35]:
!mkdir ex_2  

mkdir: cannot create directory ‘ex_2’: File exists


In [38]:
!cp H.psml  O.psml  water.fdf ex_2/ 

In [39]:
%cd ex_2

/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm/ex_2


In [41]:
Energy_Shift = [0.01, 0.001, 0.0001]

for ES in Energy_Shift:
    newf = f"ES_{ES}"

    # create directory
    !mkdir -p {newf}

    # copy input files
    !cp H.psml O.psml water.fdf {newf}/

    # modify EnergyShift in fdf
    !sed -i "s/PAO.EnergyShift.*/PAO.EnergyShift {ES} Ry/" {newf}/water.fdf

    # run siesta
    !cd {newf} && siesta water.fdf > water.out


Job completed
Job completed
Job completed


# BASIS ENTHALPY

In [44]:
 Energy_Shift = [0.01, 0.001, 0.0001]

for ES in Energy_Shift:
    print(f"ES = {ES}", end="  ")
    !grep -F "E + p_basis" ES_{ES}/water.out


ES = 0.01      (Free)E + p_basis*V_orbitals       =        -482.064557 eV
ES = 0.001      (Free)E + p_basis*V_orbitals       =        -482.435979 eV
ES = 0.0001      (Free)E + p_basis*V_orbitals       =        -482.344446 eV


In [60]:
Energy_Shift = [0.01, 0.001, 0.0001]

for ES in Energy_Shift:
    print(f"ES = {ES}", end="  ")
    !cat ES_{ES}/BASIS_ENTHALPY
 

ES = 0.01    -482.06455744736570     
 The above number is the electronic (free)energy:  -482.13276932064997     
 Plus the pressure :    1.3595723686972974E-006  (    2.0000000000000000E-002  GPa)
      times the orbital volume (in Bohr**3):    3687.5420025311787     
ES = 0.001    -482.43597920980130     
 The above number is the electronic (free)energy:  -482.58803395722634     
 Plus the pressure :    1.3595723686972974E-006  (    2.0000000000000000E-002  GPa)
      times the orbital volume (in Bohr**3):    8220.0977750288184     
ES = 0.0001    -482.34444599134150     
 The above number is the electronic (free)energy:  -482.62496318922348     
 Plus the pressure :    1.3595723686972974E-006  (    2.0000000000000000E-002  GPa)
      times the orbital volume (in Bohr**3):    15164.793163095555     


In [51]:
!pwd

/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm/ex_2/ES_0.0001


# TOTAL ENERGY

In [52]:
%cd /home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm/ex_2

/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm/ex_2


In [55]:
Energy_Shift = [0.01, 0.001, 0.0001]

for ES in Energy_Shift:
    print(f"ES = {ES}", end="  ")
    !grep -F "Total =" ES_{ES}/water.out | tail -1


ES = 0.01  siesta:         Total =    -482.132769
ES = 0.001  siesta:         Total =    -482.588034
ES = 0.0001  siesta:         Total =    -482.624963


# Total Energy vs Basis Enthalpy(Mathematically , E_opt= E+ p-basis*V_Orbitals)


ES = 0.01  siesta:    Total =    -482.132769       (Free) =    -482.064557 eV    Difference = -0.068211
ES = 0.001  siesta:   Total =    -482.588034       (Free) =    -482.435979 eV    Difference = -0.152055
ES = 0.0001  siesta:  Total =    -482.624963       (Free) =    -482.344446 eV    Difference = -0.280517
       

CHECK THE FINGDING IS CORRECT OR HAVING ERROR

In [45]:
%cd ..

/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm


In [46]:
%cd ex_2

/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm/ex_2


In [47]:
%cd ES_0.0001

/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm/ex_2/ES_0.0001


In [48]:
!ls

0_NORMAL_EXIT		     h2o.bib	      h2o.ORB_INDX    O.ion
BASIS_ENTHALPY		     h2o.BONDS	      h2o.STRUCT_OUT  O.ion.nc
BASIS_HARRIS_ENTHALPY	     h2o.BONDS_FINAL  h2o.XV	      O.ion.xml
CLOCK			     h2o.DM	      H.ion	      O.psml
fdf.20260106T153921.216.log  h2o.EIG	      H.ion.nc	      OUTVARS.yml
FORCE_STRESS		     h2o.FA	      H.ion.xml       PARALLEL_DIST
h2o.alloc		     h2o.HSX	      H.psml	      water.fdf
h2o.BASIS_ENTHALPY	     h2o.KP	      MESSAGES	      water.out


In [50]:
!grep -F "E + p_basis" water.out

    (Free)E + p_basis*V_orbitals       =        -482.344446 eV


In [62]:
!cat h2o.BASIS_ENTHALPY

 Basis enthalpy [eV] :   -482.34444599134150     
 Basis Harris enthalpy [eV] :   -482.34444599043258     
 Free energy [eV] :   -482.62496318922348     
 Harris energy [eV] :   -482.62496318831455     
 Orbital volume [Ang**3] :    2247.1904994083420     
 Average basis pressure [eV/Ang**3] :    1.2483018148921527E-004
 Average basis pressure [GPa] :    2.0000000000000000E-002


E + p_basis value for ES=0.001 has the largest negative ,so it has minimum kohn-sham total energy among all other. ES= 0.001Ry is balanced basis


# Exercise-1.3 Change the SplitNorm and check the Enthalpy, Total Energy and Orbital volume

In [63]:
%cd /home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm/

/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm


In [64]:
!mkdir ex_3


In [65]:
!cp H.psml  O.psml  water.fdf ex_3/

In [66]:
!ls ex_3

H.psml	O.psml	water.fdf


In [68]:
%cd ex_3

/home/l-rishiraj/siesta-docs/work-files/tutorials/basic/basis-optimization/01-Eshift_SplitNorm/ex_3


In [79]:
splitnorm = [0.45, 0.55, 0.65]
EnergyShift = 0.001  # Ry (fixed)

for SN in splitnorm:
    folder = f"SN_{SN}"

    # create folder
    !mkdir -p {folder}

    # copy input files
    !cp water.fdf O.psml H.psml {folder}/

    # modify SplitNorm (dimensionless!)
    !sed -i "s/PAO.SplitNorm.*/PAO.SplitNorm {SN}/" {folder}/water.fdf

    # fix EnergyShift
    !sed -i "s/PAO.EnergyShift.*/PAO.EnergyShift {EnergyShift} Ry/" {folder}/water.fdf

    # run siesta
    !cd {folder} && siesta water.fdf > water.out


Job completed
Job completed
Job completed


In [87]:
for SN in splitnorm:
    print(f"SN _{SN}", end="  ")
    !grep -F "E + p_basis" SN_{SN}*/water.out

SN _0.45      (Free)E + p_basis*V_orbitals       =        -482.696472 eV
SN _0.55      (Free)E + p_basis*V_orbitals       =        -482.741379 eV
SN _0.65      (Free)E + p_basis*V_orbitals       =        -482.621525 eV


In [88]:
for SN in splitnorm:
    print(f"SN_{SN}",end =" ")
    !grep -F "Total =" SN_{SN}*/water.out

SN_0.45 siesta:         Total =    -482.837963
SN_0.55 siesta:         Total =    -482.881867
SN_0.65 siesta:         Total =    -482.761377


In [95]:
for SN in splitnorm:
    print(f"SN_{SN}", end=" ")
    !cat SN_{SN}/*BASIS_ENTHALPY


SN_0.45   -482.69647204025256     
 The above number is the electronic (free)energy:  -482.83796294242586     
 Plus the pressure :    1.3595723686972974E-006  (    2.0000000000000000E-002  GPa)
      times the orbital volume (in Bohr**3):    7649.0150412159492     
 Basis enthalpy [eV] :   -482.69647204025256     
 Basis Harris enthalpy [eV] :   -482.69647204265760     
 Free energy [eV] :   -482.83796294242586     
 Harris energy [eV] :   -482.83796294483096     
 Orbital volume [Ang**3] :    1133.4670869288188     
 Average basis pressure [eV/Ang**3] :    1.2483018148921527E-004
 Average basis pressure [GPa] :    2.0000000000000000E-002
SN_0.55   -482.74137892246574     
 The above number is the electronic (free)energy:  -482.88186717216308     
 Plus the pressure :    1.3595723686972974E-006  (    2.0000000000000000E-002  GPa)
      times the orbital volume (in Bohr**3):    7594.8115288186527     
 Basis enthalpy [eV] :   -482.74137892246574     
 Basis Harris enthalpy [eV] :   -48

 SN_0.45  Orbital volume [Ang**3] :    1133.4670869288188  
 SN_0.55  Orbital volume [Ang**3] :    1125.4349550834477  
 SN_0.65  Orbital volume [Ang**3] :    1120.3374442047921 

# THANK YOU